### ISSUE TITLE
Run locale-dependent tests with fixed locale

### ISSUE THREAD

**Issue by flying-sheep** [kovidgoyal/rapydscript-ng#133]
Else this happens:

```
Failed running: /tmp/str.pyj-es5.js
str.pyj:
assert.js:86
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: '30,000' == '30.000'
    at test (/tmp/str.pyj-es5.js:3997:13)
    at /tmp/str.pyj-es5.js:4055:9
    at /tmp/str.pyj-es5.js:4196:7
    at /tmp/str.pyj-es5.js:4197:3
    at Script.runInContext (vm.js:135:20)
    at Script.runInNewContext (vm.js:141:17)
    at Object.runInNewContext (vm.js:307:38)
    at ~/.cache/yay/rapydscript-ng-git/src/rapydscript-ng/tools/test.js:78:20
    at Array.forEach (<anonymous>)
    at module.exports (~/.cache/yay/rapydscript-ng-git/src/rapydscript-ng/tools/test.js:45:11)


str.pyj:        test failed

There were 1 test failure(s):
str.pyj
```

**Comment by kovidgoyal** [comment]
I'm afraid I am not familiar enough with Node to know what heuristics it
uses to detect locale. Most likely running the tests with

LANG=en_US.UTF-8

should do the trick.

**Comment by flying-sheep** [comment]
Exactly. The tests should be fixed to run that code with that locale.

**Comment by eli-schwartz** [comment]
The only locale guaranteed to exist is the C locale.

C.UTF-8 would be nice, but it's been debated in circles for years now: https://sourceware.org/bugzilla/show_bug.cgi?id=17318

Setting en_US.UTF-8 in the build environment would break for non-English people who don't have one.

**Comment by kovidgoyal** [comment]
> Exactly. The tests should be fixed to run that code with that locale.

The locale has to be set globally, it cant be set for individual tests, short of major refactorings to the test running system, which is not something I think is worth doing. If you are building a package of RS then dimply set the locale in your build script before running the test suite. Other than that why does it matter to you?

**Comment by flying-sheep** [comment]
I’m trying to use @eli-schwartz python3 version of calibre, which depends on an AUR package of the rapydscript-ng project and runs its test suite.

sure, we can run the test suite there in the correct locale, but tests shouldn’t depend on some environment factor to run correctly, so that shouldn’t be necessary and the problem should be fixed here.

> The locale has to be set globally, it cant be set for individual tests, short of major refactorings to the test running system

I think setting it locally to have all of the tests run in the same locale should be possible.

> C.UTF-8 would be nice, but it's been debated in circles for years now

C.UTF-8 exists in e.g. ubuntu and fedora, but not in arch. I thing a good solution would be to set it to C.UTF-8 here to make the tests run in most distributions, but allow to override it (like `RAPYDSCRIPT_TEST_LOCALE=en_US.UTF8 bin/rapydscript test`)

That would result in the same behavior for now in arch (we have to override the locale to make the tests run), but it would lead to less hassle for people who have a system with C.UTF-8 support.

**Comment by kovidgoyal** [comment]
On Mon, Apr 29, 2019 at 12:53:33AM -0700, Philipp A. wrote:
> I’m trying to use @eli-schwartz python3 version of calibre, which depends on an AUR package of the rapydscript-ng project and runs its test suite.
> 
> sure, we can run the test suite there in the correct locale, but tests shouldn’t depend on some environment factor to run correctly, so that shouldn’t be necessary and the problem should be fixed here.

All tests depend on the environment to run correctly. Hardcoding a
locale seems weird to me. That just means the test now depend on the
system having that locale installed. It makes more sense to me to have
the person running the tests run them in whatever locale is appropriate
and present on the system. 

> 
> > The locale has to be set globally, it cant be set for individual tests, short of major refactorings to the test running system
> 
> I think setting it locally to have all of the tests run in the same locale should be possible.

Having all tests run in the same locale is what I meant by setting it
globally, and as far as I know, there is no way to do that inn Node,
short of setting the env var and re-execing.

**Comment by flying-sheep** [comment]
> That just means the test now depend on the system having that locale installed.

The tests *currently* depend on `en_US.UTF-8` being not only installed, but also the active locale. My proposal would only reduce those two assumptions to one.

And C and C.UTF-8 are built-in, so they don’t need to be installed. Sadly C.UTF-8 isn’t available everywhere, and C makes a lot of software switch off unicode support.

> Having all tests run in the same locale is what I meant by setting it globally, and as far as I know, there is no way to do that in Node, short of setting the env var and re-execing.

Exactly. Should be easy enough. (Also that’s some insane design decision, as is usual in the JS world)

**Comment by kovidgoyal** [comment]
On Mon, Apr 29, 2019 at 09:20:03AM +0000, Philipp A. wrote:
> > That just means the test now depend on the system having that locale installed.
> 
> The tests *currently* depend on `en_US.UTF-8` being not only installed, but also the active locale. My proposal would only reduce those two assumptions to one.

No, they depend on the active locale using , as the thousands separator
and . as the decimal, lots of locales meet that property. By hardcoding
one locale, this actually makes the tests more brittle.

If you really want to fix this, the right way is to edit the tests to
have them use local dependent JS APIs to generate th strings to check
against. See str.pyj

**Comment by flying-sheep** [comment]
that’s actually a good point.

**Comment by eli-schwartz** [comment]
Neat, thanks for fixing this properly @flying-sheep. ;)

BTW you're wrong about Arch "waiting for glibc upstream to support it". While it is true that we won't support glibc until upstream does, that is the fallback reason. The primary reason is because, *surprise*, Debian and Fedora use differing, incompatible implementations that produce different results and IIRC have essentially invalid definitions anyway.

**Comment by flying-sheep** [comment]
Heh. We’re now talking about two examples of why it’s good to do things properly :wink:

My fix here led to the discovery of #138, and Arch’s refusal to hack in a third way to do C.UTF-8 might help developing a proper solution.

**Comment by eli-schwartz** [comment]
Great!

The other solution is that I uploaded the non-git version of rapydscript-ng to the community repository. My current community/calibre package can use it anyway, since I rebuild the pyj in order to insert the new etags when rebuilding mathjax with --system-mathjax.
